x/or/da/keeper/cc.go

```diff
@@ -1,6 +1,326 @@
-	panic("implement me")
+import (
+	"bytes"
+	"compress/zlib"
+	"crypto/sha256"
+	"encoding/binary"
+	"io"
+	"time"
+
+	sdktypes "github.com/Finschia/finschia-sdk/types"
+	sdkerror "github.com/Finschia/finschia-sdk/types/errors"
+	"github.com/Finschia/finschia-sdk/x/or/da/types"
+)
+
+func (k Keeper) SaveQueueTx(ctx sdktypes.Context, rollupName string, tx []byte, gasLimit, L1ToL2GasRatio uint64) error {
+	// Transactions submitted to the queue lack a method for paying gas fees to the Sequencer.
+	// For transaction with a high L2 gas limit, we burn some extra gas on L1.
+	gasToConsume := (gasLimit - k.EnqueueL2GasPrepaid(ctx, tx)) / L1ToL2GasRatio
+	ctx.GasMeter().ConsumeGas(gasToConsume, "enqueue tx")
+
+	var queueState *types.QueueTxState
+	queueState, _ = k.GetQueueTxState(ctx, rollupName)
+	if queueState == nil {
+		queueState = &types.QueueTxState{
+			ProcessedQueueIndex: 0,
+			NextQueueIndex:      1,
+		}
+	}
+
+	qtx := &types.L1ToL2Queue{
+		Timestamp: ctx.BlockTime(),
+		L1Height:  ctx.BlockHeight(),
+		Txraw:     tx,
+		Status:    types.QUEUE_TX_PENDING,
+	}
+
+	k.saveQueueTx(ctx, rollupName, queueState.NextQueueIndex, qtx)
+	queueState.NextQueueIndex++
+
+	k.setQueueTxState(ctx, rollupName, queueState)
+
+	if err := ctx.EventManager().EmitTypedEvent(&types.EventSaveQueueTx{
+		RollupName:       rollupName,
+		NextQueueIndex:   queueState.NextQueueIndex,
+		ExtraConsumedGas: gasToConsume,
+		L2GasLimit:       gasLimit,
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (k Keeper) SaveCCBatch(ctx sdktypes.Context, rollupName string, batch *types.CCBatch) error {
+	var ccState *types.CCState
+	ccState, _ = k.GetCCState(ctx, rollupName)
+	if ccState == nil {
+		if batch.ShouldStartAtFrame != 0 {
+			return types.ErrInvalidCCBatch.Wrapf("cc state not found: this batch should start at frame 0, but start at %d", batch.ShouldStartAtFrame)
+		}
+		ccState = &types.CCState{
+			Base:             1,
+			Height:           0,
+			ProcessedL2Block: 0,
+			Timestamp:        time.Unix(0, 0).UTC(),
+			L1Height:         0,
+		}
+	}
+
+	var queueState *types.QueueTxState
+	queueState, _ = k.GetQueueTxState(ctx, rollupName)
+	if queueState == nil {
+		queueState = &types.QueueTxState{
+			ProcessedQueueIndex: 0,
+			NextQueueIndex:      1,
+		}
+	}
+
+	var totalFrames uint64
+	if ccState.Height == 0 {
+		totalFrames = uint64(len(batch.Frames))
+	} else {
+		prevCCRef, err := k.GetCCRef(ctx, rollupName, ccState.Height)
+		if err != nil {
+			return err
+		}
+
+		if prevCCRef.TotalFrames != batch.ShouldStartAtFrame {
+			return types.ErrInvalidCCBatch.Wrapf("batch should start at frame %d but this batch start at frame %d", prevCCRef.TotalFrames, batch.ShouldStartAtFrame)
+		}
+		totalFrames = prevCCRef.TotalFrames + uint64(len(batch.Frames))
+	}
+
+	batchHash := sha256.Sum256(k.cdc.MustMarshal(batch))
+	ref := types.NewCCRef(ctx.TxBytes(), uint32(ctx.MsgIndex()), uint32(len(batch.Frames)), totalFrames, batchHash[:])
```

Comment:
Shouldn't it be converted from TxBytes to txhash?
```suggestion
	ref := types.NewCCRef(sha256.Sum256(ctx.TxBytes()), uint32(ctx.MsgIndex()), uint32(len(batch.Frames)), totalFrames, batchHash[:])
```

Reply:
`sha256.Sum256(ctx.TxBytes())` is txhash

Reply:
Yes, it seems correct to give the CCRef a txhash.
```
func NewCCRef(txhash []byte, msgIdx, batchSize uint32, totalFrames uint64, batchRoot []byte) *CCRef {
	return &CCRef{
		TxHash:      txhash,
		MsgIndex:    msgIdx,
		TotalFrames: totalFrames,
		BatchSize:   batchSize,
		BatchRoot:   batchRoot,
	}
}
```